src/videojs-contrib-hls.js

```diff
@@ -207,124 +114,41 @@ export default class HlsHandler extends Component {
     } else if (videojs.options.hls) {
       this.options_.withCredentials = videojs.options.hls.withCredentials;
     }
-    this.playlists = new Hls.PlaylistLoader(this.source_.src,
-                                            this.options_.withCredentials);
-
-    this.tech_.one('canplay', this.setupFirstPlay.bind(this));
-
-    this.playlists.on('loadedmetadata', () => {
-      oldMediaPlaylist = this.playlists.media();
-
-      // if this isn't a live video and preload permits, start
-      // downloading segments
-      if (oldMediaPlaylist.endList &&
-          this.tech_.preload() !== 'metadata' &&
-          this.tech_.preload() !== 'none') {
-        this.loadingState_ = 'segments';
-        this.segments.playlist(this.playlists.media());
-        this.segments.load();
-      }
-
-      this.setupSourceBuffer_();
-      this.setupFirstPlay();
-      this.tech_.trigger('loadedmetadata');
-    });
-
-    this.playlists.on('error', () => {
-      this.blacklistCurrentPlaylist_(this.playlists.error);
-    });
-
-    this.playlists.on('loadedplaylist', () => {
-      let updatedPlaylist = this.playlists.media();
-      let seekable;
-
-      if (!updatedPlaylist) {
-        // select the initial variant
-        this.playlists.media(this.selectPlaylist());
-        return;
-      }
-
-      this.segments.playlist(updatedPlaylist);
-      this.updateDuration(this.playlists.media());
 
-      // update seekable
-      seekable = this.seekable();
-      if (this.duration() === Infinity &&
-          seekable.length !== 0) {
-        this.mediaSource.addSeekableRange_(seekable.start(0), seekable.end(0));
-      }
-
-      oldMediaPlaylist = updatedPlaylist;
+    this.masterPlaylistController_ = new MasterPlaylistController({
+      url: this.source_.src,
+      withCredentials: this.options_.withCredentials,
+      currentTimeFunc: this.tech_.currentTime.bind(this.tech_),
+      mediaSourceMode: this.mode_,
+      hlsHandler: this,
+      externHls: Hls
     });
 
-    this.playlists.on('mediachanging', () => {
-      this.segments.pause();
-    });
-    this.playlists.on('mediachange', () => {
-      this.segments.abort();
-      this.segments.load();
-      this.tech_.trigger({
-        type: 'mediachange',
-        bubbles: true
-      });
-    });
-
-    this.segments = new SegmentLoader({
-      currentTime: this.tech_.currentTime.bind(this.tech_),
-      mediaSource: this.mediaSource,
-      withCredentials: this.options_.withCredentials
-    });
-
-    this.segments.on('progress', () => {
-      // figure out what stream the next segment should be downloaded from
-      // with the updated bandwidth information
-      this.bandwidth = this.segments.bandwidth;
-      this.playlists.media(this.selectPlaylist());
-
-      this.tech_.trigger('progress');
-    });
-    this.segments.on('error', () => {
-      this.blacklistCurrentPlaylist_(this.segments.error());
-    });
+    this.tech_.one('canplay',
```

Comment:
`canplay` indicates [media data is available in the buffer](https://html.spec.whatwg.org/multipage/embedded-content.html#event-media-canplay). That should never happen until we begin buffering and we don't begin buffering live streams until you hit play. Seems like this condition shouldn't be hit.